[Zsigmond] iP

zsiggg opened the pull request:
### Zuke ###
> "I spent too much time on this" - Unknown

Zuke helps you remember your tasks. It 
- remembers the task list even after you close it
- is really simple to use on command line!

Why would you use Zuke?
1. You love the command-line
2. You need a simple task manager without too many extra features that apps like Notion provide
3. You're ~a geek~ _cool_ 🤓

How do you start using Zuke?
- [ ] Run the .jar file. Download it [here](https://github.com/zsiggg/ip/releases/tag/A-Jar).
- [ ] Add a task with /todo, /event, /deadline.
- [ ] Mark or unmark a task when you are done with the task.
- [ ] Close the program with /bye.

Zuke is object-oriented, with classes like `Command` and `Parser` one of the many used to encapsulate its functions. Here is the `main` method, found in the `Duke` class:
```java
public static void main(String[] args) {
    new Duke("data.txt", "data").run();
}
```

Commit:
Add Gradle support

zsiggg committed:
feat: greet, echo, exit

zsiggg committed:
style: print lines before and after output

zsiggg committed:
feat: add, list

zsiggg committed:
refactor: task class

zsiggg committed:
fix: allow whitespace in task

zsiggg committed:
refactor: mark and unmark in task class

zsiggg committed:
feat: mark

zsiggg committed:
feat: unmark

zsiggg committed:
refactor: todo, deadline, event classes

zsiggg committed:
feat: add todo, event, deadlines

zsiggg committed:
refactor: default task type will be todo; task class will be abstract

zsiggg committed:
feat: automated text ui testing

zsiggg committed:
feat: error checking

must prefix with todo, event, deadline to add task; unrecognised text
will print an error
marking and unmarking error -> not a number provide; index out of range
todo error -> no text
deadline/event error -> no text; no time
lsit error -> no tasks in list

zsiggg committed:
feat: delete

zsiggg committed:
feat: output add and delete commands into file

zsiggg committed:
feat: read previous tasks on start

added isPrinting booleans, so that previous commands read from file will
not cause outputs to be printed

zsiggg committed:
feat: output marking and unmarking into file; /done option when adding tasks

zsiggg committed:
refactor: ignore files in data folder

zsiggg committed:
fix: cross platform file paths

zsiggg committed:
feat: store deadline and event time as LocalDateTime, and output prettily

zsiggg committed:
Merge branch 'branch-level-8'

zsiggg committed:
style: no indentation for case

zsiggg committed:
refactor: use enums for switch case

made fields for output folder, name, path private since only used in
this class
took out the nested switch case in the outer switch case's default case

zsiggg committed:
refactor: encapsulate code for adding into respective classes

needed a constructor that took in the command string and parsed it to
get the relevant fields. however, super() needs to be called at the top
of the constructor.

thus, created factory methods for each subclass of Task, and made
constructors private.

zsiggg committed:
fix: terminate upon bye command

zsiggg committed:
refactor: tasklist class

zsiggg committed:
refactor: parser, command classes

previously used switch case within Duke to handle different commands.

encapsulated the switch case inside the Parser class, which returns a
Command to the caller or throw an exception if the command is not valid. execute() method can then be called to execute the command.

zsiggg committed:
refactor: storage class

zsiggg committed:
refactor: taskList nullary constructor, used in Storage.load()

zsiggg committed:
refactor: ui class

zsiggg committed:
refactor: clean up Duke class to make use of new classes

zsiggg committed:
refactor: command classes

need to differentiate between when to print and not to print (i.e.
during loading from output file) in execute() in command classes. thus
overloaded execute() method and called the one that does not take in Ui
in storage.load().

DeleteCommand class now stores Task, such that we have access to the
Task after it is removed from the TaskList to call the toString()
method.

program does not exit gracefully upon bye command due to system.exit(1)
being used. thus, added isExit boolean in Duke.run() and ByeCommand, exiting Duke.run() on isExit = true.

zsiggg committed:
refactor: remove TaskList(String) constructor

zsiggg committed:
refactor: showWelcome() and printWithDivider()

add newline to end of string to be printed in printWithDivider().

corrected showWelcome() message.

zsiggg committed:
refactor: platform-independent file paths

take in directories of the output file path as a variable argument in
Duke() and Storage.load(), then join them with the correct separator
according to the platform in use to form the paths.

zsiggg committed:
refactor: remove isPrinting in Task subclasses

isPrinting was used to control whether the acknowledge message is
printed when a task is created. however, since this should be done in Ui
class, this should be removed.

zsiggg committed:
refactor: ui class

print new line in printWithDivider(), so that there will be a spacing
between two consecutive outputs of this function.

showLoadingError() for use if storage.load() fails in Duke class.

zsiggg committed:
fix: getTaskIndex() should return 0-based index

output of getTaskIndex() is used directly in UpdateCommand subclasses
get and mark/unmark a task.

however, the user passes in a 1-based index,
but the taskList methods takes in 0-based indexes.

thus it is more accurate for getTaskIndex() to return a 0-based index,
that can then be used directly in taskList methods.

zsiggg committed:
refactor: storage class

instead of in Duke class, initialise FileWriter class, write() and
close() in Storage class instead, since Storage should take care of all
interactions with the output file.

write() to be used to write changes after the TaskList is changed by a
command.

list tasks with ListCommand upon loading of tasks in load() completes.

zsiggg committed:
refactor: write changes to taskList into output file

in TaskList class, every method that results in change to the tasks will
now write to the output file, by calling storage.write(). thus, the
output file is updated whenever there are changes to the tasks due to a
command. when catching the IOException, changes made to tasks and
addCommands are reversed, since we are unable to save this change to the
output file, then the IOException is rethrown with a message.

in Command class, execute() should take in storage, which will be passed
into command.execute(), then to the methods in TaskList. execute() will
only throw IOException if no Ui is passed in, as there is no way to
print a message to inform the user that an error has occured.

in TaskList class, mark() and unmark() added, which were initially only
in Task class. this is these cause an update to the TaskList, which
should be updated with storage.write(). in toString(), streams are used
instead of concatenating, since it is more efficient.

zsiggg committed:
fix: catch exceptions in Duke class

IllegalArgumentException thrown by Parser.parse should be caught and
their messages printed, as this denotes a wrong command being entered.
otherwise, this would cause the program to be terminated upon a wrong
command.

catch IOException now thrown by storage.load(), and assign a new empty
instance of TaskList. this is to prevent a half-loaded TaskList from
being used, in the event that the IOException was thrown when some lines
had already been read from the output file.

zsiggg committed:
refactor: divide classes into packages (duke, duke.command, duke.task)

zsiggg committed:
Merge remote-tracking branch 'public/add-gradle-support'

zsiggg committed:
test: Parser.parse (wrong command), tasklist.add, ui.showLoadingError

zsiggg committed:
doc: javadoc comments for all classes

zsiggg committed:
docs: javadoc comments for all public methods

zsiggg committed:
refactor: remove unused TaskList parameter in Parser.getTaskIndex()

zsiggg committed:
feat: find command

execute() in FindCommand uses a new method search() in TaskList, which
returns a TaskList instance representing the matching tasks. search()
filters the tasks and addCommands with the condition that the
corresponding task's text field contains the keyword. search()
uses a new private TaskList constructor that takes in predefined tasks
and addCommands lists.

ongwenpin commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,159 @@
+        }
+    }
+
+    public Task get(int i) throws IndexOutOfBoundsException {
```
Perhaps its better to name the method more explicitly like getTask so there is less confusion in the future where they might be similar get methods?

ongwenpin commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,83 @@
+package duke;
+
+import duke.command.*;
```
Perhaps it is better to list out explicitly what classes you have imported from package?

zsiggg replied:
You're right, I should have listed out all of the commands explicitly for easier readability and future maintenance.

ongwenpin reviewed (commented):
Overall, code looks good and very well organised, LGTM!

zsiggg committed:
refactor: rename get() to getTask() in TaskList

zsiggg committed:
refactor: list imports instead of using wildcard in Parser

zsiggg committed:
feat: Checkstyle

zsiggg committed:
style: fix checkstyle violations in duke.command

zsiggg committed:
style: fix checkstyle violations in duke.task

zsiggg committed:
style: fix checkstyle violations in duke

zsiggg committed:
style: fix checkstyle violations for tests

zsiggg committed:
feat: gui that repeats command

zupey commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,158 @@
+        }
+    }
+
+    public TaskList search(String keyword) {
```
Remember to add javadoc comments for the methods that you add! 😄

zsiggg replied:
Resolved in a future branch!

zupey commented on src/test/java/duke/ParserTest.java:
```diff
@@ -0,0 +1,17 @@
+
+public class ParserTest {
+    @Test
+    public void parseWrongCommandTest() {
```
```suggestion
    public void parse_wrongCommand_exceptionThrown() {
```
Perhaps you can use this naming convention to better indicate the function you are testing!

zsiggg replied:
Thanks for pointing out the test method naming convention :)

zupey commented on src/test/java/duke/TaskListTest.java:
```diff
@@ -0,0 +1,31 @@
+public class TaskListTest {
+    private final TaskList taskList = new TaskList();
+    @Test
+    public void addTest() {
```
```suggestion
    public void add_deadline_successfullyAddedToTaskList() {
```
Can consider renaming test methods to be more descriptive 😄

zsiggg replied:
Done!

zupey commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,158 @@
+        ArrayList<Task> matchingTasks = new ArrayList<>();
+        ArrayList<String> matchingAddCommands = new ArrayList<>();
+
+        IntStream.range(0, tasks.size())
```
I like that you also used functional programming in an overall OOP project! 👍

zupey reviewed (commented):
LGTM, except for some small things that you may have missed, such as not including javadoc.

Liked that you used streams and all the best for the iP and upcoming tP.

zsiggg committed:
feat: gui

added gui on top of the existing cli. run gui with Launcher.java, run
Cli with Duke.java

refactored Ui to an abstract class, with Gui and Cli extending it,
taking care of message printing for the different interfaces.

zsiggg committed:
refactor: 3 part test method naming convention

zsiggg committed:
Merge branch 'branch-level-10'

zsiggg committed:
Merge branch 'branch-A-Checkstyle'

zsiggg committed:
refactor: add assertions

Classes in the command package accept any string as the command when
they are initialised.

This could result in a wrong command being created if the wrong
constructor is called for a certain command string (e.g. a DeleteCommand
could possibly be created for "todo abc", which should be an
AddTodoCommand).

Assertions are added to all factory methods and constructors that assert
what their command strings should start with or be equal to.

Since Parser should already ensure that the correct constructors and
factory methods are called for a certain command string, asserts ensure
that this does not change in future updates in code.

zsiggg committed:
refactor: improved code quality

Constructors and factory methods of the command package violates some
code quality points.

Need to improve code quality to increase readability.

Let's,
* rename variables to better describe what they are
* shift code out of else block to ensure that happy path is as unindented as possible
* add whitespace between different logical groups of code

zsiggg committed:
feat: github actions for CI

zsiggg committed:
style: fix checkstyle violations, mainly for new classes added for gui

zsiggg committed:
chore: remove i/o redirection tests for github actions

zsiggg committed:
refactor: change sad face emoji to :(, as not mappable in windows

zsiggg committed:
test: remove UiTest

zsiggg committed:
Merge branch 'master' into branch-A-Assertions

zsiggg committed:
Merge pull request #2 from zsiggg/branch-A-Assertions

refactor: add assertions

zsiggg committed:
Merge branch 'master' into branch-A-CodeQuality

zsiggg committed:
Merge pull request #3 from zsiggg/branch-A-CodeQuality

Code Quality

zsiggg committed:
feat: add ability to tag when adding task

words after # and before any other flags (e.g. /by) will be taken as a
tag. currently, only one tag is supported per task.

zsiggg committed:
fix: edit welcome message

zsiggg committed:
refactor: pretty GUI

window can be expanded without changes to the look of the GUI.

change profile pictures from default.

fixed issue of long messages not being showed in the dialog box.

changed text bubbles to cover the whole row instead.

different coloured text bubbles for Duke and user.

zsiggg committed:
docs: add screenshot of GUI

zsiggg committed:
docs: add README

zsiggg committed:
fix: correct mainClassName in build.gradle for GUI

zsiggg committed:
docs: fix emojis not showing in README
